### PR TITLE
Add descriptive identifiers to blog post links

### LINKED ISSUE #155: Posts do not have titles or descriptive links
## Problem

* Posts currently have no dedicated titles; the first H1 heading in the post body is implied to be the title. This limits any further use of H1 headings.

    ~In contrast, many websites have a dedicated title field, rendering them using the `<title>` HTML element and freeing up H1 headings for post sections.~ The `<title>` element is metadata; websites often use the `<h1>` element (or other custom elements) for page titles. See [Adding a title (MDN Web Docs)](https://developer.mozilla.org/en-US/docs/Learn/HTML/Introduction_to_HTML/The_head_metadata_in_HTML#adding_a_title).

    This issue is most relevant to blog posts.

* Blog posts currently have non-descriptive links in the format `/blog/<36-character-post-id>`. Visitors cannot quickly understand what a post is about, or whether they are visiting the right post at all, just by examining its link.

## Suggestions

1. ~Add a separate title field (e.g., `post/title`) in the data and validation schemas~ Verify that the user's post contains exactly 1 H1 heading, and then use this as the title

    See [Heading levels (MDN Web Docs writing style guide)](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#heading_levels).

3. Use the title as part of the post link, e.g., `/blog/<post-id>/<post_title_in_snake_case>`

    - Shorten the `<post-id>` part from 36 characters to 7 or 8

### PR DESCRIPTION
**Pull request by flybot-nam-nguyenhoai**
## Closes #155

- [x] New posts are checked to ensure their Markdown content contains exactly one H1 heading, and that this heading appears at the very beginning. This is then considered the post title.[^1]
- [x] Post URLs now include descriptive identifiers (also called *slugs*[^2]) that reflect their corresponding post titles.
    - [x] Test that the new descriptive URL actually leads to the correct post.

[^1]: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#heading_levels
[^2]: https://developer.mozilla.org/en-US/docs/Glossary/Slug

### CODE REVIEW

File: common/src/flybot/common/utils.cljc [hunk 1]
```diff
@@ -1,5 +1,9 @@
 (ns flybot.common.utils
-  (:require #?(:clj [datalevin.core :as d])))
+  #?@
+   (:clj
+    [(:require [datalevin.core :as d] [markdown-to-hiccup.core :as mth])]
+    :cljs
+    [(:require [markdown-to-hiccup.core :as mth])]))
```
Comment (skydread1) [inline common/src/flybot/common/utils.cljc:6]: Can be simplified like this:

```clojure
(ns flybot.common.utils
  (:require [markdown-to-hiccup.core :as mth]
            #?(:clj [datalevin.core :as d])))
```

File: client/web/src/flybot/client/web/core/dom/common.cljs [hunk 2]
```diff
@@ -22,3 +27,80 @@
...
+                                     :post/page
+                                     (= page-name))))
+         matches-id-ending? (fn [id] (str/ends-with? (str id) id-ending))
+         matches-url-identifier? (fn [post] (= url-identifier (mk-post-url-identifier post)))
+         queried-posts (->> @(rf/subscribe [:subs/pattern {:app/posts '?x}])
+                            (utils/filter-map-kv
+                             (fn [[id post]]
+                               (and (matches-id-ending? id)
+                                    (matches-page? post)
+                                    (matches-url-identifier? post)))))]
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/common.cljs:105]: can use `into` directly here for clarity with the arity `(into to xform from)`

```clojure
queried-posts (into
                        {}
                        (filter
                         (fn [[id post]]
                           (and (matches-id-ending? id)
                                (matches-page? post)
                                (matches-url-identifier? post))))
                        @(rf/subscribe [:subs/pattern {:app/posts '?x}]))
```

File: common/src/flybot/common/utils.cljc [hunk 3]
```diff
@@ -24,3 +28,40 @@
...
+  Probably expensive. Use sparingly."
+  [pred map]
+  (into {} (filter #(pred (key %)) map)))
+
+(defn filter-vals
+  "Returns map entries for which (pred key) returns logical true.
+
+  Probably expensive. Use sparingly."
+  [pred map]
+  (into {} (filter #(pred (val %)) map)))
```
Comment (skydread1) [inline common/src/flybot/common/utils.cljc:51]: unused so can remove

File: client/web/src/flybot/client/web/core/dom/common.cljs [hunk 4]
```diff
@@ -22,3 +27,80 @@
...
+      "Untitled_post")))
+
+(defn get-post-by-id-page
+  "Retrieves a post (or posts) with the given page name, ID ending and URL
+  identifier (slug).
+
+  Returns the post as an {id post} map. If multiple matches are found, they are
+  all included in the map. If no matches are found, returns an empty map."
+  ([page-name id-ending url-identifier]
+   (let [matches-page? (fn [post] (or (= :all page-name)
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/common.cljs:94]: What is the purpose of `(= :all page-name)`?

**Review by skydread1 — CHANGES_REQUESTED** [review]
In general we want to have as few functions as possibles in `utils`.

Same goes for the too generic `common` ns.

- There is a `hiccup` namespace already existing where you can move your hiccup related helper functions
- In the common folder, you can create `link.cljs` to add your internal-link logic
- Maybe some post related helper functions should be added in post.cljs directly or a subfolder for post.

You need to identify where your logic is needed (who requires it) and you should be able to get rid of the common.cljs namespace which is bad practice

File: client/web/src/flybot/client/web/core/dom/common.cljs (hunk 2 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/common.cljs:105]: Today I learned about transducers. Thanks!

File: client/web/src/flybot/client/web/core/dom/common.cljs (hunk 4 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/common.cljs:94]: I initially thought I might need it to visit a post when I wouldn't be sure what page it belonged to.

Removed.

File: common/src/flybot/common/utils.cljc (hunk 1 above)
Comment (flybot-nam-nguyenhoai) [reply common/src/flybot/common/utils.cljc]: Reverted to only requiring `datalevin.core`.

File: client/web/src/flybot/client/web/core/dom/common.cljs (hunk 2 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/common.cljs]: Fixed.

File: client/web/src/flybot/client/web/core/dom/page.cljs [hunk 5]
```diff
@@ -30,23 +49,47 @@
...
+                         {:parameters
+                          {:path
+                           {:id-ending '?id-ending
+                            :url-identifier '?url-identifier}}}}])
+        queried-post (-> (get-post-by-id-page :blog
+                                              query-id-ending
+                                              query-url-identifier)
+                         vals
+                         first
+                         add-post-hiccup-content)]
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/page.cljs:92]: good!

File: common/src/flybot/common/validation/markdown.cljc [hunk 6]
```diff
@@ -0,0 +1,17 @@
+(ns flybot.common.validation.markdown
+  (:require [markdown-to-hiccup.core :as mth]))
+
+(defn has-valid-h1-title?
+  "Checks that the given Markdown content contains exactly one H1 heading at
+  the start."
+  [^String md-content]
```
Comment (skydread1) [inline common/src/flybot/common/validation/markdown.cljc:7]: You don't have to type hint here.

Our type validations are all made via malli.

In general in Clojure, we use type hint as last resort for compilation optimisation for instance which we don't need in this case.

**Review by skydread1 — CHANGES_REQUESTED** [review]
Some clj tests are failing which is expected as you modified a malli schema which is used both both frontend and backend validation.

Good reflex is to run the clj locally before pushing.

File: common/src/flybot/common/validation/markdown.cljc (hunk 6 above)
Comment (flybot-nam-nguyenhoai) [reply common/src/flybot/common/validation/markdown.cljc:7]: Got it. I actually don't remember why I put the type hint there.

File: client/web/src/flybot/client/web/core/dom/page.cljs (hunk 5 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page.cljs:92]: Thank you! (I'm not sure what I did well, but thank you anyway.)

**Comment by flybot-nam-nguyenhoai** [comment]
To test the new descriptive URLs, I'm thinking of doing the following:

1. `(test-fixtures)` to initialize
2. Create a new post
3. Create a descriptive link using `(post-link post)`
4. Open the link (somehow)
5. Check that the new `app-db` page contains only the new post

Steps 4 and 5 seem tricky at the moment, so any help will be greatly appreciated.

**Comment by skydread1** [comment]
> To test the new descriptive URLs, I'm thinking of doing the following:
> 
> 1. `(test-fixtures)` to initialize
> 2. Create a new post
> 3. Create a descriptive link using `(post-link post)`
> 4. Open the link (somehow)
> 5. Check that the new `app-db` page contains only the new post
> 
> Steps 4 and 5 seem tricky at the moment, so any help will be greatly appreciated.

We can focus on checking if the `:current-view` has been properly updated (it would be updated when you call `internal-link` because it will trigger the dispatch from `on-navigate`:
```clojure
(rf/dispatch [:evt.page/set-current-view new-match])
``` 

A simple test for the `apply` page (without any params) could be:
```clojure
(deftest internal-link
  (rf-test/run-test-sync
   (test-fixtures)
   (sut/internal-link :flybot/apply "Apply")
   (let [current-view (rf/subscribe [:subs/pattern
                                     {:app/current-view
                                      {:data
                                       {:name '?
                                        :page-name '?}
                                       :path-params '?}}
                                     :all])]
     (testing "Navigates to apply page."
       (is (= {:app/current-view 
               {:data {:name :flybot/home, :page-name :home}
                :path-params {}}}
            (get @current-view '&?)))))))
```

File: common/src/flybot/common/validation/markdown.cljc (hunk 6 above)
Comment (flybot-nam-nguyenhoai) [reply common/src/flybot/common/validation/markdown.cljc]: They got in the way during testing, too.

Removed.

**Review by skydread1 — APPROVED** [review]

**Pull request merged**